vision/use-cases/video-thumbnail-generation/video_thumbnail_generation.ipynb

```diff
@@ -0,0 +1,804 @@
+        "    )\n",
+        "    # calling Gemini API\n",
+        "    responses = model.generate_content(\n",
+        "        [video_input, prompt],\n",
+        "        generation_config=generation_config,\n",
+        "        stream=False,\n",
+        "    )\n",
+        "    # print(responses.text, end=\"\")\n",
+        "    return responses.text"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "18d7efd1dfe2"
+      },
+      "source": [
+        "### Defining a function to parse output and display results"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 9,
+      "metadata": {
+        "id": "e6f4c6427c3a"
+      },
+      "outputs": [],
+      "source": [
+        "def display_results(\n",
+        "    response_text: str,\n",
+        "    clip: moviepy.video.io.VideoFileClip.VideoFileClip,\n",
+        ") -> None:\n",
+        "    \"\"\"Parse json output, extract thumbnail frames and plot.\"\"\"\n",
+        "    # cleaning up and loading json object\n",
+        "    response_string = re.sub(\"```json\", \"\", response_text)\n",
+        "    response_string = re.sub(\"```\", \"\", response_string)\n",
+        "    json_response = json.loads(response_string)\n",
+        "\n",
+        "    # extracting timestamps and captions from response\n",
+        "    timestamps = json_response[\"timestamp\"]\n",
+        "    captions = json_response[\"caption\"]\n",
+        "\n",
+        "    # Image plotting settings\n",
+        "    fig, ax = plt.subplots(nrows=2, ncols=2, figsize=(12, 9))\n",
+        "\n",
+        "    # extract frame for each timestamp and plot the images\n",
+        "    counter = 0\n",
+        "    for timestamp, caption in zip(timestamps, captions):\n",
+        "        frame = clip.get_frame(timestamp)\n",
+        "        row, col = counter // 2, counter % 2\n",
+        "        ax[row, col].imshow(frame)\n",
+        "        ax[row, col].set_title(caption, fontdict={\"fontsize\": 9})\n",
+        "        counter += 1\n",
+        "\n",
+        "    fig.show()"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "0d39180622a0"
+      },
+      "source": [
+        "# Case 1: Using a Simple Prompt"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "5bb3ff574d22"
+      },
+      "source": [
+        "### Writing a basic prompt for thumbnail generation"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 10,
+      "metadata": {
+        "id": "e96a4213e6d3"
+      },
+      "outputs": [],
+      "source": [
+        "basic_prompt = \"\"\"Generate 4 thumbnail images from the given video file with short captions.\n",
+        "Provide response in proper JSON format similar to following:\n",
+        "{\n",
+        "    \\\"timestamp\\\": [],\n",
+        "    \\\"caption\\\": []\n",
+        "}\n",
```

Comment:
Use Controlled Generation https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/control-generated-output to get proper JSON output instead of prompting.

Reply:
Done